Cryo pod UI

Fruitsalad opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

***Revision 2***

## About the PR
This PR adds a new window for cryo pods. See the media section.

Relevant issue: #41704

### Notes on revision 2

The new version of this PR has the health analyzer and the cryo info side-by-side. @Princess-Cheeseballs requested this because in the future the health analyzer UI will be more horizontal. I've added a checklist and some flavor text on the cryo side of the UI, mainly to fill up vertical space (although the checklist is pretty useful as well). If necessary these can also be removed in the future to make space for new UI sections (I was told there'd be some additions for auto-resuscitation).

## Why / Balance
Small note concerning gameplay balance: I turned the cryo pod's injection speed down to inject 1u every two seconds, so that it matches the metabolism rate of cryo medicine. This is intended to help you waste less cryoxadone.

## Technical details
<details>
<summary>HealthAnalyzerControl, HealthAnalyzerWindow</summary>

- I copied the body of `HealthAnalyzerWindow` into `HealthAnalyzerControl`, which is reused by the cryo pod UI. The health analyzer window is now just a thin wrapper.
- Nothing was changed here besides inserting the struct `HealthAnalyzerUiState` (which is used for both `HealthAnalyzerUserMessage` and `CryoPodUserMessage`).

</details>
<details>
<summary>HealthAnalyzerSystem, HealthAnalyzerScannedUserMessage</summary>

- Split `HealthAnalyzerUiState` from `HealthAnalyzerScannedUserMessage`. Same data, but now reusable for `CryoPodUserMessage`.
- Split off the function `GetHealthAnalyzerUiState`, which generates a `HealthAnalyzerUiState` based on the health of the given entity.

</details>
<details>
<summary>GasAnalyzerSystem</summary>

- I needed `GasMixEntry` to fill in `CryoPodUserMessage`, so I split off the function `GenerateGasMixEntry`.

</details>
<details>
<summary>CryoPodWindow</summary>

- This is the new UI.
- Populate: Most of this function seems fairly straightforward to myself, but the temperature warning code is somewhat complex. It looks through all the chemicals in the beaker for the one with the lowest temperature requirement and displays a warning if the gas temperature is higher than that.
- MeasureOverride: When the patient is gasping and getting oxygen damage and that damage is also getting healed within a second, that was pretty annoying for a while, because the buttons in the UI would jiggle up and down. To remedy this, the UI only shrinks in height when the difference is at least 80.

</details>
<details>
<summary>BeakerBarChart</summary>

- This is the segmented bar chart on CryoPodWindow that displays the available chemicals.
- The API is based on SplitBar. BeakerBarChart adds animations and a ruler.
- BeakerBarChart also fixes a bug that SplitBar still has where tooltips would flash every time the entries are updated.
- BeakerBarChart keeps persistent entries. `Clear()` doesn't actually remove entries but instead indicates that they should start animating to 0u. Entries are kept alive after `Clear()` by calling `SetEntry` before the BeakerBarChart gets a chance to update.

</details>
<details>
<summary>CryoPodSystem, SharedCryoPodSystem, CryoPodComponent, cryo_pod.yml</summary>

- Cryo pods no longer have a HealthAnalyzerComponent. It seems to me this was a hacky way to make the old health analyzer UI work.
- Added `CryoPodUserMessage`, which transmits the data needed for `CryoPodWindow`.
- `CryoPodSystem` now sends a `CryoPodUserMessage` once per second (of course only for `ActiveCryoPodComponents`)
- Moved medicine-injection code from `SharedCryoPodSystem` into the serverside `CryoPodSystem` to reduce duplication with the UI update code.
- Cryo pods now have two solution containers: The old beaker is still there, and a new `InjectionBuffer` was added so that you don't end up inadvertently injecting the whole beaker into the patient, as used to be the case.
- When the patient is removed from the cryo pod the injection buffer is cleared and chemicals inside are lost. The gameplay reason for this is to prevent people from filling the injection buffer with poison that you can't easily get rid of.
- `CryoPodSystem` also handles the UI's inject & eject button press events.
- You can now open the cryo pod UI when the pod is empty.

</details>

## Media

<img width="604" height="415" alt="with_patient" src="https://github.com/user-attachments/assets/c4b88257-63bf-4317-9aa6-04b1b24db5be" />

<details>
<summary>More screenshots and videos</summary>


<img width="270" height="468" alt="not_set_up" src="https://github.com/user-attachments/assets/11aa1a8f-5816-4742-a085-11f1160772f7" />

At roundstart

<img width="302" height="403" alt="set_up" src="https://github.com/user-attachments/assets/8003c982-8964-4614-a719-35cc2b469c15" />

After setting it up

https://github.com/user-attachments/assets/0237c09f-2bdb-4105-af67-f028adea312b

Healing a patient

https://github.com/user-attachments/assets/9906a562-f6d5-4158-865d-90b57c69c96e

Setting it up (slightly chaotic video, sorry)

</details>

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an in-game showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: Added a new user interface for cryo pods

YotaXP commented:
Did a little testing. I've found that swapping out the beaker (or ejecting the patient) while it's injecting completely empties the injection buffer, destroying whatever reagents were in it. There are reasons we want to be able to swap beakers during the healing process, such as administering leporazine just before ejection, so this could be an unpleasant surprise.

It would be lovely to have an "eject beaker" ⏏️ button next to the chem bar, if you can fit that.

Fruitsalad commented:
I was aware of those issues but I figured they were minor enough to ignore. I've fixed them now though. For the mob temperature, I've changed it so the status says "Cooling patient..." when they're not at temperature yet. I think showing a warning is too severe since there's nothing wrong.

I wasn't sure whether I should make it say "cooling" or "chilling" or "freezing". Let me know what you think.

<img width="586" height="553" alt="cooling" src="https://github.com/user-attachments/assets/0565b455-944a-47c3-9b2c-4f3a83c819ba" />

https://github.com/user-attachments/assets/2d55d3a6-aff6-4cb7-8fb0-d95786d43efd

Princess-Cheeseballs reviewed (approved):
Looks good. I'm gonna have someone more familiar with UI code look over it and once they approve this should be merged.

sowelipililimute commented:
UI bugs: the custom bar control doesn't work at >100% UI scales
<img width="289" height="132" alt="Bildschirmfoto 2025-12-25 um 4 41 20 PM" src="https://github.com/user-attachments/assets/acb73b41-9989-4edd-b915-16db1eeada9e" />

the constant recreation of the gas bar also means that the tooltip repeatedly disappears and re-appears when you try to view it

the behaviour of the chemical gauge gets goofy with progressively larger beakers
<img width="295" height="131" alt="Bildschirmfoto 2025-12-25 um 4 47 50 PM" src="https://github.com/user-attachments/assets/1978c110-5eaf-491b-845c-ce537c26429f" />

UI design issues:
<img width="241" height="57" alt="Bildschirmfoto 2025-12-25 um 4 42 56 PM" src="https://github.com/user-attachments/assets/4ddc2a26-8779-46aa-b15c-3d59077a5802" />

this element is a bit overly fluff and doesn't need to be there given the existence of the titlebar

wording nits:
<img width="179" height="40" alt="Bildschirmfoto 2025-12-25 um 4 48 32 PM" src="https://github.com/user-attachments/assets/87a934fe-8558-4b14-8900-fe4e40b20fb6" />
this can be phrased as "READY TO INJECT" 

<img width="276" height="102" alt="Bildschirmfoto 2025-12-25 um 4 49 48 PM" src="https://github.com/user-attachments/assets/d181516c-ebe4-4c5d-b35c-773d8b747291" />

the error states for these should be rephrasings and not just `-- NO`, i.e. `No chemicals available`

good error states should also ideally point towards how you resolve them, so the UI should read something like

- **Underpressurized**
  Insert gas into the cryo loop
- **No chemicals available**
  Insert a beaker with chemicals
- **Above cryogenic temperatures**
  Turn on the freezer

sowelipililimute commented on Content.Client/HealthAnalyzer/UI/HealthAnalyzerControl.xaml.cs:
```diff
@@ -0,0 +1,243 @@
+// Health analyzer UI is split from its window because it's used by both the
+// health analyzer item and the cryo pod UI.
+
+namespace Content.Client.HealthAnalyzer.UI
```
style nit: `namepsace Content.Client.HealthAnalyzer.UI;` and drop the brackets

Princess-Cheeseballs commented:
> this element is a bit overly fluff and doesn't need to be there given the existence of the titlebar

I prefer it being there over it just being empty space personally.

sowelipililimute commented on Content.Client/Medical/Cryogenics/BeakerBarChart.cs:
```diff
@@ -0,0 +1,268 @@
+        {
+            var targetWidthFraction = entry.TargetAmount / Capacity;
+
+            if (entry.WidthFraction == targetWidthFraction)
```
don't disable the float equality comment you should be comparing their difference against some epsilon instead of waiting for it to maybe converge on a bit-for-bit equal number

`(entity.WidthFraction - targetWidthFraction) < SOME_EPSILON`

Fruitsalad replied:
This one's actually fine already: The `abs(WidthFraction - targetWidthFraction) < epsilon` is just a couple of lines down. First the function skips any entries that are already exactly the target width, for the rest it calculates the new width and if the new width is close enough to the target width, it sets the width to exactly the target width. Exact comparison is fine here.

I've done a lot of geometry algorithm stuff which is extremely error-prone if you aren't deliberate about your fuzzy float comparisons, so I prefer to be very deliberate with exact & fuzzy comparisons.

sowelipililimute commented on Content.Shared/Medical/Cryogenics/CryoPodComponent.cs:
```diff
@@ -96,3 +118,48 @@ public enum CryoPodVisuals : byte
+        Type = type;
+        Quantity = quantity;
+    }
+}
```
use separate message types, not an enum + nullable fields

Fruitsalad commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -6,61 +6,61 @@
-        // TODO: This should be a state my dude
-        _uiSystem.ServerSendUiMessage(
+        var patient = entity.Comp.BodyContainer.ContainedEntity;
+        var gasMix = _gasAnalyzerSystem.GenerateGasMixEntry("Cryo pod", air.Air);
```
I've skipped this one because this string is not actually visible. In the gas analyzer UI you have section headers where this string would show up, but the cryo pod UI doesn't show it. I'm only giving it a value for the sake of completeness.

Fruitsalad commented:
Besides the stuff I've already commented on, I fixed the bugs (except the gas bar flickering tooltip bug) and replaced "Nominal" with "Ready to inject", as well as adding the status messages "Injecting..." and "Ready for patient".

With regards to the wording nits: I meant for the UI to look somewhat like industrial equipment, so I would personally prefer to not make it too user-friendly/handholding. I added the `Pressurized — NO` phrasing specifically to make it look more industrial/Linux-ish and less user-friendly/Mac-ish. It's not a hill I'll die on but I would prefer to keep the phrasings less user-friendly.

For the checklist descriptions, I think the checklist items as they are, combined with the warnings, are probably enough guidance for someone who can figure out how to play SS14, but I'm open to adding the descriptions if it's a strong preference. I am however very limited in typography by the available style classes. I haven't tried it yet but I suspect it would be difficult to make the checklist descriptions look good.

> the constant recreation of the gas bar also means that the tooltip repeatedly disappears and re-appears when you try to view it

This is an inherent issue with SplitBar — I very briefly mentioned it in the OP. The gas analyzer UI also has this flickering tooltip problem. I fixed it for BeakerBarChart, whose API is based on SplitBar, but I'd prefer to leave the SplitBar issue out-of-scope. I'm of the opinion that a rework of SplitBar should be a separate PR since it also affects the gas analyzer UI.
